utxonursery_test.go

```diff
@@ -379,3 +387,708 @@ func TestBabyOutputSerialization(t *testing.T) {
+	}
+
+	if onLocalCommitment {
+		timeoutTx := &wire.MsgTx{
+			TxIn: []*wire.TxIn{
+				{
+					PreviousOutPoint: htlcOp,
+					Witness:          [][]byte{{}},
+				},
+			},
+			TxOut: []*wire.TxOut{
+				{},
+			},
+		}
+
+		outgoingRes.SignedTimeoutTx = timeoutTx
+	} else {
+		outgoingRes.ClaimOutpoint = htlcOp
+	}
+
+	return &outgoingRes
+}
+
+func createCommitmentRes() *lnwallet.CommitOutputResolution {
+	// Set up a commitment output resolution to hand off to nursery.
+	commitRes := lnwallet.CommitOutputResolution{
+		SelfOutPoint: wire.OutPoint{},
+		SelfOutputSignDesc: lnwallet.SignDescriptor{
+			Output: &wire.TxOut{
+				Value: 10000,
+			},
+		},
+		MaturityDelay: 2,
+	}
+
+	return &commitRes
+}
+
+func incubateTestOutput(t *testing.T, nursery *utxoNursery,
+	onLocalCommitment bool) *lnwallet.OutgoingHtlcResolution {
+
+	outgoingRes := createOutgoingRes(onLocalCommitment)
+
+	// Hand off to nursery.
+	err := nursery.IncubateOutputs(
+		testChanPoint,
+		nil,
+		[]lnwallet.OutgoingHtlcResolution{*outgoingRes},
+		nil,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// IncubateOutputs is executing synchronously and we expect the output
+	// to immediately show up in the report.
+	expectedStage := uint32(2)
+	if onLocalCommitment {
+		expectedStage = 1
+	}
+
+	// TODO(joostjager): Nursery is currently not reporting this limbo
+	// balance.
+	if onLocalCommitment {
+		assertNurseryReport(t, nursery, 1, expectedStage, 10000)
+	}
+
+	return outgoingRes
+}
+
+func assertNurseryReport(t *testing.T, nursery *utxoNursery,
+	expectedNofHtlcs int, expectedStage uint32,
+	expectedLimboBalance btcutil.Amount) {
+	report, err := nursery.NurseryReport(&testChanPoint)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(report.htlcs) != expectedNofHtlcs {
+		t.Fatalf("expected %v outputs to be reported, but report "+
+			"only contains %v", expectedNofHtlcs, len(report.htlcs))
+	}
+
+	if expectedNofHtlcs != 0 {
+		htlcReport := report.htlcs[0]
+		if htlcReport.stage != expectedStage {
+			t.Fatalf("expected htlc be advanced to stage %v, but "+
+				"it is reported in stage %v",
+				expectedStage, htlcReport.stage)
+		}
+	}
+
+	if report.limboBalance != expectedLimboBalance {
+		t.Fatalf("expected limbo balance to be %v, but it is %v instead",
+			expectedLimboBalance, report.limboBalance)
+	}
+}
+
+func assertNurseryReportUnavailable(t *testing.T, nursery *utxoNursery) {
+	_, err := nursery.NurseryReport(&testChanPoint)
+	if err != ErrContractNotFound {
+		t.Fatal("expected report to be unavailable")
+	}
+}
+
+// testRestartLoop runs the specified test multiple times and in every run it
+// will attempt to execute a restart action in a different location. This is to
+// assert that the unit under test is recovering correctly from restarts.
+func testRestartLoop(t *testing.T, test func(*testing.T,
+	func(func()) bool)) {
+
+	// Start with running the test without any restarts (index zero)
+	restartIdx := 0
+
+	for {
+		currentStartStopIdx := 0
```

Comment:
It is indeed a little hard to parse, but this was the way to keep the amount of duplicated code in the tests minimal.

What you describe (running the test twice) is not what is happening. 

The test indeed first runs without restart. I updated the code to actually use `no_restart` as subtest name to make that more clear.

After that, the test is ran `n` more times. `n` is the number of times a call to `ctx.Restart()` is made in the test code. Example output:
```
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/no_restart
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/restart_1
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/restart_2
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/restart_3
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/restart_4
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/restart_5
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/restart_6
--- PASS: TestNurseryOutgoingHtlcSuccessOnRemote (1.01s)
```
So 6+1 times in this case.

It will determine on the fly how many calls to `ctx.Restart()` are present in the code. So if you want to add an extra restart point, there is nothing more to do than just add a `ctx.Restart()` call. The actual restart code that is then executed, is specified in `createNurseryTestContext` (stop/start nursery). Running the test `n` times, keeping counters and detecting the final run, is done in `testRestartLoop`.